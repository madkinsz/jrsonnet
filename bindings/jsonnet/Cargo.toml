[package]
name = "jsonnet"
description = "Rust implementation of libjsonnet.so"
version = "0.4.2"
authors = ["Yaroslav Bolyukin <iam@lach.pw>"]
license = "MIT"
edition = "2021"
publish = false

[dependencies]
jrsonnet-evaluator = { path = "../../crates/jrsonnet-evaluator", version = "0.4.2" }
jrsonnet-parser = { path = "../../crates/jrsonnet-parser", version = "0.4.2" }
jrsonnet-gcmodule = { version = "0.3.4" }

[lib]
crate-type = ["cdylib"]

[features]
interop = []
experimental = ["exp-preserve-order", "exp-destruct"]
exp-preserve-order = ["jrsonnet-evaluator/exp-preserve-order"]
exp-destruct = ["jrsonnet-evaluator/exp-destruct"]
