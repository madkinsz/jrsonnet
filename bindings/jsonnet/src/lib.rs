#[cfg(feature = "interop")]
pub mod interop;

pub mod import;
pub mod native;
pub mod val_extract;
pub mod val_make;
pub mod val_modify;
pub mod vars_tlas;

use std::{
	alloc::Layout,
	ffi::{CStr, CString},
	os::raw::{c_char, c_double, c_int, c_uint},
	path::PathBuf,
};

use import::NativeImportResolver;
use jrsonnet_evaluator::{IStr, ManifestFormat, State, Val};

/// WASM stub
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn _start() {}

#[no_mangle]
pub extern "C" fn jsonnet_version() -> &'static [u8; 8] {
	b"v0.16.0\0"
}

#[no_mangle]
pub extern "C" fn jsonnet_make() -> *mut State {
	let state = State::default();
	state.with_stdlib();
	state.settings_mut().import_resolver = Box::new(NativeImportResolver::default());
	Box::into_raw(Box::new(state))
}

/// # Safety
#[no_mangle]
#[allow(clippy::boxed_local)]
pub unsafe extern "C" fn jsonnet_destroy(vm: *mut State) {
	Box::from_raw(vm);
}

#[no_mangle]
pub extern "C" fn jsonnet_max_stack(vm: &State, v: c_uint) {
	vm.settings_mut().max_stack = v as usize;
}

// jrsonnet currently have no GC, so these functions is no-op
#[no_mangle]
pub extern "C" fn jsonnet_gc_min_objects(_vm: &State, _v: c_uint) {}
#[no_mangle]
pub extern "C" fn jsonnet_gc_growth_trigger(_vm: &State, _v: c_double) {}

#[no_mangle]
pub extern "C" fn jsonnet_string_output(vm: &State, v: c_int) {
	match v {
		1 => vm.set_manifest_format(ManifestFormat::String),
		0 => vm.set_manifest_format(ManifestFormat::Json {
			padding: 4,
			#[cfg(feature = "exp-preserve-order")]
			preserve_order: false,
		}),
		_ => panic!("incorrect output format"),
	}
}

/// # Safety
///
/// This function is most definitely broken, but it works somehow, see TODO inside
#[no_mangle]
pub unsafe extern "C" fn jsonnet_realloc(_vm: &State, buf: *mut u8, sz: usize) -> *mut u8 {
	if buf.is_null() {
		assert!(sz != 0);
		return std::alloc::alloc(Layout::from_size_align(sz, std::mem::align_of::<u8>()).unwrap());
	}
	// TODO: Somehow store size of allocation, because its real size is probally not 16 :D
	// OR (Alternative way of fixing this TODO)
	// TODO: Standard allocator uses malloc, and it doesn't uses allocation size,
	// TODO: so it should work in normal cases. Maybe force allocator for this library?
	let old_layout = Layout::from_size_align(16, std::mem::align_of::<u8>()).unwrap();
	if sz == 0 {
		std::alloc::dealloc(buf, old_layout);
		return std::ptr::null_mut();
	}
	std::alloc::realloc(buf, old_layout, sz)
}

/// # Safety
#[no_mangle]
#[allow(clippy::boxed_local)]
pub unsafe extern "C" fn jsonnet_json_destroy(_vm: &State, v: *mut Val) {
	Box::from_raw(v);
}

#[no_mangle]
pub extern "C" fn jsonnet_max_trace(vm: &State, v: c_uint) {
	vm.set_max_trace(v as usize)
}

/// # Safety
///
/// This function is safe, if received v is a pointer to normal C string
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_file(
	vm: &State,
	filename: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	match vm
		.import(PathBuf::from(filename.to_str().unwrap()))
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest(v))
	{
		Ok(v) => {
			*error = 0;
			CString::new(&*v as &str).unwrap().into_raw()
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}

/// # Safety
///
/// This function is safe, if received v is a pointer to normal C string
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_snippet(
	vm: &State,
	filename: *const c_char,
	snippet: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	let snippet = CStr::from_ptr(snippet);
	match vm
		.evaluate_snippet(
			filename.to_str().unwrap().into(),
			snippet.to_str().unwrap().into(),
		)
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest(v))
	{
		Ok(v) => {
			*error = 0;
			CString::new(&*v as &str).unwrap().into_raw()
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}

fn multi_to_raw(multi: Vec<(IStr, IStr)>) -> *const c_char {
	let mut out = Vec::new();
	for (i, (k, v)) in multi.iter().enumerate() {
		if i != 0 {
			out.push(0);
		}
		out.extend_from_slice(k.as_bytes());
		out.push(0);
		out.extend_from_slice(v.as_bytes());
	}
	out.push(0);
	out.push(0);
	let v = out.as_ptr();
	std::mem::forget(out);
	v as *const c_char
}

/// # Safety
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_file_multi(
	vm: &State,
	filename: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	match vm
		.import(PathBuf::from(filename.to_str().unwrap()))
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest_multi(v))
	{
		Ok(v) => {
			*error = 0;
			multi_to_raw(v)
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}

/// # Safety
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_snippet_multi(
	vm: &State,
	filename: *const c_char,
	snippet: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	let snippet = CStr::from_ptr(snippet);
	match vm
		.evaluate_snippet(
			filename.to_str().unwrap().into(),
			snippet.to_str().unwrap().into(),
		)
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest_multi(v))
	{
		Ok(v) => {
			*error = 0;
			multi_to_raw(v)
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}

fn stream_to_raw(multi: Vec<IStr>) -> *const c_char {
	let mut out = Vec::new();
	for (i, v) in multi.iter().enumerate() {
		if i != 0 {
			out.push(0);
		}
		out.extend_from_slice(v.as_bytes());
	}
	out.push(0);
	out.push(0);
	let v = out.as_ptr();
	std::mem::forget(out);
	v as *const c_char
}

/// # Safety
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_file_stream(
	vm: &State,
	filename: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	match vm
		.import(PathBuf::from(filename.to_str().unwrap()))
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest_stream(v))
	{
		Ok(v) => {
			*error = 0;
			stream_to_raw(v)
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}

/// # Safety
#[no_mangle]
pub unsafe extern "C" fn jsonnet_evaluate_snippet_stream(
	vm: &State,
	filename: *const c_char,
	snippet: *const c_char,
	error: &mut c_int,
) -> *const c_char {
	let filename = CStr::from_ptr(filename);
	let snippet = CStr::from_ptr(snippet);
	match vm
		.evaluate_snippet(
			filename.to_str().unwrap().into(),
			snippet.to_str().unwrap().into(),
		)
		.and_then(|v| vm.with_tla(v))
		.and_then(|v| vm.manifest_stream(v))
	{
		Ok(v) => {
			*error = 0;
			stream_to_raw(v)
		}
		Err(e) => {
			*error = 1;
			let out = vm.stringify_err(&e);
			CString::new(&out as &str).unwrap().into_raw()
		}
	}
}
