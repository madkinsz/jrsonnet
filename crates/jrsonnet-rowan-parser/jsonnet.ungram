// This file describes structure of jsonnet source code
// It is also used to generate files in src/generated

// Token names ending with `!` are considered meta, and handled specifically

SourceFile = Expr

ExprBinary =
    lhs:LhsExpr
    BinaryOperator
    rhs:Expr
ExprUnary =
    UnaryOperator
    rhs:Expr
ExprSlice =
    Expr
    SliceDesc
ExprIndex =
    Expr
    '.'
    index:Name
ExprIndexExpr =
    base:LhsExpr
    '['
    index:Expr
    ']'
ExprApply =
    Expr
    ArgsDesc
    'tailstrict'?
ExprObjExtend =
    LhsExpr
    Expr
ExprParened =
    '('
    Expr
    ')'

ExprLiteral =
    Literal
ExprIntrinsicThisFile =
    '$intrinsicThisFile'
ExprIntrinsicId =
    '$intrinsicId'
ExprIntrinsic =
    '$intrinsic'
    '('
    name:Name
    ')'
ExprString =
    Text
ExprNumber =
    Number
ExprArray =
    '['
    (Expr (',' Expr)* ','?)?
    ']'
ExprObject =
    '{'
    ObjBody
    '}'
ExprArrayComp =
    '['
    Expr
    ','?
    CompSpec*
    ']'

ExprImport =
    ImportKind Text

ImportKind =
    'importstr'
|   'importbin'
|   'import'

ExprVar =
    name:Name
ExprLocal =
    'local'
    (Bind (',' Bind)* ','?)
    ';'
    Expr

ExprIfThenElse =
    'if'
    cond:Expr
    'then'
    then:TrueExpr
    ('else' else_:FalseExpr)?

ExprFunction =
    'function'
    '('
    ParamsDesc
    ')'
    Expr
ExprAssert =
    Assertion
    ';'
    Expr
ExprError =
    'error'
    Expr

Expr =
    ExprBinary
|   ExprUnary
|   ExprSlice
|   ExprIndex
|   ExprIndexExpr
|   ExprApply
|   ExprObjExtend
|   ExprParened
|   ExprIntrinsicThisFile
|   ExprIntrinsicId
|   ExprIntrinsic
|   ExprString
|   ExprNumber
|   ExprLiteral
|   ExprArray
|   ExprObject
|   ExprArrayComp
|   ExprImport
|   ExprVar
|   ExprLocal
|   ExprIfThenElse
|   ExprFunction
|   ExprAssert
|   ExprError

BinaryOperator =
    '||' | '&&'
|   '|' | '^' | '&'
|   '==' | '!=' | '<' | '>' | '<=' | '>=' | 'in'
|   '<<' | '>>'
|   '+' | '-'
|   '*' | '/' | '%'
|   'ERROR_NO_OPERATOR!'

UnaryOperator =
    '-' | '!' | '~'

SliceDescEnd=Expr
SliceDescStep=Expr
SliceDesc =
    '['
    from:Expr?
    ':'
    (
        end:SliceDescEnd?
        (
            ':'
            step:SliceDescStep?
        )?
    )?
    ']'

Name =
    'LIT_IDENT!'

ArgsDesc =
    '('
    (Arg (',' Arg)* ','?)?
    ')'
Arg =
    (name:Name '=')? Expr

ObjBodyComp =
    pre:ObjLocalPostComma*
    '['
    key:LhsExpr
    ']'
    '+'?
    ':'
    value:Expr
    post:ObjLocalPreComma*
    CompSpec*
ObjBodyMemberList =
    (Member (',' Member)* ','?)?
ObjBody =
    ObjBodyComp
|   ObjBodyMemberList

ObjLocalPostComma =
    ObjLocal
    ','
ObjLocalPreComma =
    ','
    ObjLocal

MemberBindStmt = ObjLocal
MemberAssertStmt = Assertion
MemberField = Field
Member =
    MemberBindStmt
|   MemberAssertStmt
|   MemberField

ObjLocal =
    'local'
    Bind

FieldNormal =
    FieldName
    '+'?
    Visibility
    Expr
FieldMethod =
    FieldName
    ParamsDesc
    Visibility
    Expr
Field =
    FieldNormal
|   FieldMethod

FieldNameFixed =
    id:Name
|   Text
FieldNameDynamic =
    '['
    Expr
    ']'
FieldName =
    FieldNameFixed
|   FieldNameDynamic

Visibility =
    ':::'
|   '::'
|   ':'

Literal =
    'null'
|   'true'
|   'false'
|   'self'
|   '$'
|   'super'

Text =
    'LIT_STRING_DOUBLE!'
|   'ERROR_STRING_DOUBLE_UNTERMINATED!'
|   'LIT_STRING_SINGLE!'
|   'ERROR_STRING_SINGLE_UNTERMINATED!'
|   'LIT_STRING_DOUBLE_VERBATIM!'
|   'ERROR_STRING_DOUBLE_VERBATIM_UNTERMINATED!'
|   'LIT_STRING_SINGLE_VERBATIM!'
|   'ERROR_STRING_SINGLE_VERBATIM_UNTERMINATED!'
|   'ERROR_STRING_VERBATIM_MISSING_QUOTES!'
|   'LIT_STRING_BLOCK!'
|   'ERROR_STRING_BLOCK_UNEXPECTED_END!'
|   'ERROR_STRING_BLOCK_MISSING_NEW_LINE!'
|   'ERROR_STRING_BLOCK_MISSING_TERMINATION!'
|   'ERROR_STRING_BLOCK_MISSING_INDENT!'

Number =
    'LIT_FLOAT!'
|   'ERROR_FLOAT_JUNK_AFTER_POINT!'
|   'ERROR_FLOAT_JUNK_AFTER_EXPONENT!'
|   'ERROR_FLOAT_JUNK_AFTER_EXPONENT_SIGN!'

ForSpec =
    'for'
    bind:Name
    'in'
    Expr
IfSpec =
    'if'
    Expr
CompSpec =
    ForSpec
|   IfSpec

BindDestruct =
    into:Destruct
    '='
    value:Expr
BindFunction =
    name:Name
    params:ParamsDesc
    '='
    value:Expr
Bind =
    BindDestruct
|   BindFunction

ParamsDesc =
    '('
    (Param (',' Param)* ','?)?
    ')'
Param =
    Destruct
    (
        '='
        Expr
    )?

Assertion =
    'assert'
    condition:LhsExpr
    (
        ':'
        message:Expr
    )?

DestructFull =
    Name
DestructSkip =
    '?'
DestructArray =
    '['
    (
        DestructArrayPart
        (',' DestructArrayPart)*
        ','?
    )?
    ']'
DestructObject =
    '{'
    (
        DestructObjectField
        (',' DestructObjectField)*
        ','?
    )?
    DestructRest?
    ','?
    '}'
Destruct =
    DestructFull
|   DestructSkip
|   DestructArray
|   DestructObject

DestructArrayElement =
    Destruct
DestructArrayPart =
    DestructArrayElement
|   DestructRest

DestructRest =
    '...'
    into:Name?

DestructObjectField =
    field:Name
    (
        ':'
        Destruct
    )?
    (
        '='
        Expr
    )?

// Aliases used to resolve node type conflicts
TrueExpr=Expr
FalseExpr=Expr
LhsExpr=Expr

// Trivia - tokens which will be implicitly skipped for parser
Trivia =
    'LIT_WHITESPACE!'
|   'LIT_MULTI_LINE_COMMENT!'
|   'ERROR_COMMENT_TOO_SHORT!'
|   'ERROR_COMMENT_UNTERMINATED!'
|   'LIT_SINGLE_LINE_HASH_COMMENT!'
|   'LIT_SINGLE_LINE_SLASH_COMMENT!'
