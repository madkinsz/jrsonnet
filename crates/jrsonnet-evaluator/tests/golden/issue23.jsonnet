import 'issue23.jsonnet'
