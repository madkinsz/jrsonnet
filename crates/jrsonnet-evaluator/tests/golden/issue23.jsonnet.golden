infinite recursion detected
   issue23.jsonnet:1:1-26: import "issue23.jsonnet"